lib/reversetunnel/localsite.go

```diff
@@ -497,10 +497,120 @@ func (s *localSite) periodicFunctions() {
 			if err := s.sshTunnelStats(); err != nil {
 				s.log.Warningf("Failed to report SSH tunnel statistics for: %v: %v.", s.domainName, err)
 			}
+			s.agentStats()
 		}
 	}
 }
 
+// agentStats leverages the accessPoint cache to report all agents/proxies/auth servers connected to the
+// teleport cluster via prometheus metrics
+func (s *localSite) agentStats() {
+	hostID := make(map[string]struct{})
+	versionCount := make(map[string]int)
+
+	// Nodes, Proxies, Auths, KubeServices, and WindowsDesktopServers use the UUID as the name field where
+	// DB and App store it in the spec. Check expiry due to DynamoDB taking up to 48hr to expire from backend
+	// and then store hostID and version count information.
+	serverCheck := func(server interface{}) {
+		type serverKubeWindows interface {
+			Expiry() time.Time
+			GetName() string
+			GetTeleportVersion() string
+		}
+		type appDB interface {
+			serverKubeWindows
+			GetHostID() string
+		}
+
+		ttl := s.clock.Now().Add(-1 * apidefaults.ServerAnnounceTTL)
+
+		// appDB needs to be first as it also matches the serverKubeWindows interface
+		switch server.(type) {
+		case appDB:
+			if server.(appDB).Expiry().Before(ttl) {
+				return
+			}
+			if _, present := hostID[server.(appDB).GetHostID()]; !present {
+				hostID[server.(appDB).GetHostID()] = struct{}{}
+				versionCount[server.(appDB).GetTeleportVersion()]++
+			}
+		case serverKubeWindows:
+			if server.(serverKubeWindows).Expiry().Before(ttl) {
+				return
+			}
+			if _, present := hostID[server.(serverKubeWindows).GetName()]; !present {
+				hostID[server.(serverKubeWindows).GetName()] = struct{}{}
+				versionCount[server.(serverKubeWindows).GetTeleportVersion()]++
+			}
+		}
+	}
+
+	proxyServers, err := s.accessPoint.GetProxies()
+	if err != nil {
+		log.Debugf("Failed to get Proxies for teleport_registered_agent metric: %v", err)
+	}
+	if err == nil {
```

Comment:
- Why not just use `else` here?
- Why does this logic exist only for `GetProxies()`? Is there some special case with this result such that you can't just unconditionally iterate over `proxyServers` and expect it to be empty in the error case?

Reply:
Oversight when changing logic around these calls and missed this one. No special logic surrounding `GetProxies()`